File: Exp_5/RSA.java
import java.math.BigInteger;
import java.nio.charset.StandardCharsets;
import java.util.Scanner;

public class RSA {

    // Function to calculate GCD using BigInteger
    public static BigInteger getGCD(BigInteger a, BigInteger b) {
        return b.equals(BigInteger.ZERO) ? a : getGCD(b, a.mod(b));
    }

    // Function to check if a number is prime
    public static boolean isPrime(BigInteger number) {
        return number.isProbablePrime(100); // Use a probabilistic primality test
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        BigInteger n = null, e = null, d = null; // Store n, e, d

        while (true) {
            System.out.println("Menu:");
            System.out.println("1. Key Generation");
            System.out.println("2. Encryption");
            System.out.println("3. Decryption");
            System.out.println("4. Exit");
            System.out.print("Choose an option: ");
            int choice = scanner.nextInt();

            switch (choice) {
                case 1: // Key Generation
                    System.out.print("Enter prime p (up to 100 digits): ");
                    BigInteger p = new BigInteger(scanner.next());

                    System.out.print("Enter prime q (up to 100 digits): ");
                    BigInteger q = new BigInteger(scanner.next());

                    if (!isPrime(p) || !isPrime(q)) {
                        System.out.println("Both p and q must be prime numbers.");
                        break;
                    }

                    n = p.multiply(q);
                    BigInteger phi = (p.subtract(BigInteger.ONE)).multiply(q.subtract(BigInteger.ONE));

                    System.out.println("Modulus n: " + n);
                    System.out.print("Enter public key e (should be coprime with phi): ");
                    e = new BigInteger(scanner.next());

                    // Validate e
                    if (getGCD(e, phi).equals(BigInteger.ONE)) {
                        // Compute d
                        d = e.modInverse(phi);
                        System.out.println("Public key (e, n): (" + e + ", " + n + ")");
                        System.out.println("Private key (d, n): (" + d + ", " + n + ")");
                    } else {
                        System.out.println("Invalid public key e. It must be coprime with phi.");
                    }
                    break;

                case 2: // Encryption
                    if (n == null || e == null) {
                        System.out.println("Please generate keys first.");
                        break;
                    }
                    System.out.print("Enter the plaintext message as a string: ");
                    scanner.nextLine(); // Consume newline
                    String message = scanner.nextLine();

                    StringBuilder cipherText = new StringBuilder();
                    for (char ch : message.toCharArray()) {
                        BigInteger messageBigInt = BigInteger.valueOf((int) ch);
                        BigInteger encryptedChar = messageBigInt.modPow(e, n);
                        cipherText.append(encryptedChar).append(" ");
                    }
                    System.out.println("Cipher text: " + cipherText.toString().trim());
                    break;

                case 3: // Decryption
                    if (n == null || d == null) {
                        System.out.println("Please generate keys first.");
                        break;
                    }
                    System.out.print("Enter the encrypted message as integers separated by spaces: ");
                    scanner.nextLine(); // Consume newline
                    String[] encryptedMessageParts = scanner.nextLine().split(" ");

                    StringBuilder decryptedMessage = new StringBuilder();
                    for (String part : encryptedMessageParts) {
                        BigInteger encryptedChar = new BigInteger(part);
                        BigInteger decryptedCharBigInt = encryptedChar.modPow(d, n);
                        char decryptedChar = (char) (decryptedCharBigInt.intValue());
                        decryptedMessage.append(decryptedChar);
                    }
                    System.out.println("Decrypted message: " + decryptedMessage.toString());
                    break;

                case 4: // Exit
                    scanner.close();
                    System.out.println("Exiting...");
                    return;

                default:
                    System.out.println("Invalid option. Please try again.");
            }
        }
    }
}

File: Exp_5/rsa.py
from math import gcd
import sys

def is_prime(number):
    """Check if a number is prime using a probabilistic primality test."""
    if number < 2:
        return False
    # Simple primality test for demonstration
    for i in range(2, int(number ** 0.5) + 1):
        if number % i == 0:
            return False
    return True

def mod_inverse(e, phi):
    """Calculate the modular multiplicative inverse of e (mod phi)."""
    def extended_gcd(a, b):
        if a == 0:
            return b, 0, 1
        gcd, x1, y1 = extended_gcd(b % a, a)
        x = y1 - (b // a) * x1
        y = x1
        return gcd, x, y

    gcd, x, _ = extended_gcd(e, phi)
    if gcd != 1:
        raise ValueError("Modular inverse does not exist")
    return x % phi

def main():
    n = None
    e = None
    d = None

    while True:
        print("\nMenu:")
        print("1. Key Generation")
        print("2. Encryption")
        print("3. Decryption")
        print("4. Exit")
        
        try:
            choice = int(input("Choose an option: "))
        except ValueError:
            print("Please enter a valid number")
            continue

        if choice == 1:  # Key Generation
            try:
                p = int(input("Enter prime p (up to 100 digits): "))
                q = int(input("Enter prime q (up to 100 digits): "))

                if not is_prime(p) or not is_prime(q):
                    print("Both p and q must be prime numbers.")
                    continue

                n = p * q
                phi = (p - 1) * (q - 1)

                print(f"Modulus n: {n}")
                e = int(input("Enter public key e (should be coprime with phi): "))

                # Validate e
                if gcd(e, phi) == 1:
                    d = mod_inverse(e, phi)
                    print(f"Public key (e, n): ({e}, {n})")
                    print(f"Private key (d, n): ({d}, {n})")
                else:
                    print("Invalid public key e. It must be coprime with phi.")
            except ValueError as ve:
                print(f"Invalid input: {ve}")

        elif choice == 2:  # Encryption
            if None in (n, e):
                print("Please generate keys first.")
                continue

            message = input("Enter the plaintext message as a string: ")
            cipher_text = []
            
            for char in message:
                # Convert character to integer and encrypt
                message_int = ord(char)
                encrypted_char = pow(message_int, e, n)  # Using built-in modular exponentiation
                cipher_text.append(str(encrypted_char))
            
            print("Cipher text:", " ".join(cipher_text))

        elif choice == 3:  # Decryption
            if None in (n, d):
                print("Please generate keys first.")
                continue

            try:
                encrypted_message = input("Enter the encrypted message as integers separated by spaces: ")
                encrypted_parts = encrypted_message.split()
                
                decrypted_message = ""
                for part in encrypted_parts:
                    encrypted_char = int(part)
                    decrypted_char_int = pow(encrypted_char, d, n)  # Using built-in modular exponentiation
                    decrypted_message += chr(decrypted_char_int)
                
                print("Decrypted message:", decrypted_message)
            except ValueError:
                print("Invalid input. Please enter space-separated integers.")

        elif choice == 4:  # Exit
            print("Exiting...")
            sys.exit(0)

        else:
            print("Invalid option. Please try again.")

if __name__ == "__main__":
    main()